File: ml/rules/detector.py
"""Deterministic fan-out -> convergence detector.

Pattern: one source account sends to several distinct intermediary accounts
within a short window (fan-out), and a common threshold of those
intermediaries then forward money to one common collector account, also
within a short window, after receiving it (convergence).

This module has no knowledge of any specific account id, label, or the demo
fixture. Every account id it ever sees comes from the transactions passed
in. Thresholds are configuration, not constants baked into the logic, so the
same code can be pointed at a different dataset without modification.

The `score` this module produces is a hand-defined heuristic combining three
observable ratios (see `Finding.score_method` and `Finding.evidence`). It is
NOT a calibrated probability of fraud, NOT a trained model's output, and NOT
a measured performance metric of anything. It is a way to rank findings by
how strongly the observed evidence matches the fan-out/convergence pattern,
nothing more. Treat it the same way you'd treat a heuristic sort key.
"""

from __future__ import annotations

import datetime
from collections import defaultdict
from dataclasses import dataclass, field
from typing import Sequence

from .transactions import Transaction, deduplicate_by_id


@dataclass(frozen=True)
class DetectorConfig:
    """Thresholds for the fan-out/convergence pattern. All are explicit and
    tunable; none are derived from any specific dataset."""

    # Minimum number of distinct intermediaries required on both sides
    # (fan-out from the source, and convergence onto the collector) for a
    # candidate group to be reported at all.
    min_intermediaries: int = 3

    # A fan-out group is a source account's outgoing transactions that fall
    # within this many seconds of the first transaction in the group.
    fan_out_window_seconds: int = 60

    # From the moment an intermediary receives a fan-out transaction, a
    # subsequent outgoing transaction from that intermediary counts as
    # convergence evidence only if it occurs within this many seconds.
    convergence_window_seconds: int = 60

    # Used only to normalize the time-compactness term of the score into
    # [0, 1]; does not gate whether a finding is reported.
    max_scoring_window_seconds: int = 120


@dataclass(frozen=True)
class Finding:
    """A network-level piece of evidence: this pattern, at these accounts,
    backed by these specific transactions."""

    pattern: str
    source_account: str
    collector_account: str
    intermediary_accounts: tuple[str, ...]
    fan_out_transaction_ids: tuple[str, ...]
    convergence_transaction_ids: tuple[str, ...]
    window_start: datetime.datetime
    window_end: datetime.datetime
    score: float
    score_method: str
    evidence: dict = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "pattern": self.pattern,
            "source_account": self.source_account,
            "collector_account": self.collector_account,
            "intermediary_accounts": list(self.intermediary_accounts),
            "fan_out_transaction_ids": list(self.fan_out_transaction_ids),
            "convergence_transaction_ids": list(self.convergence_transaction_ids),
            "window_start": self.window_start.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "window_end": self.window_end.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "score": self.score,
            "score_method": self.score_method,
            "evidence": self.evidence,
            "score_is_not_a_probability": True,
        }


_SCORE_METHOD = (
    "heuristic_v1 = 0.5*intermediary_ratio + 0.3*amount_conservation + 0.2*time_compactness, "
    "clipped to [0, 1]. A hand-defined evidence-strength ranking score, not a calibrated "
    "probability, not a trained model output, and not a measured performance metric."
)


def _fan_out_groups(source: str, outgoing: list[Transaction], config: DetectorConfig):
    """Yield candidate fan-out groups for one source account.

    `outgoing` must already be sorted by timestamp. Each group is anchored at
    one starting transaction and extends forward while still inside the
    fan-out window, so overlapping windows starting at different transactions
    are all considered as separate candidates (deduplicated later).
    """
    n = len(outgoing)
    for i in range(n):
        window_end_time = outgoing[i].timestamp + datetime.timedelta(seconds=config.fan_out_window_seconds)
        group: list[Transaction] = [outgoing[i]]
        for j in range(i + 1, n):
            if outgoing[j].timestamp > window_end_time:
                break
            if outgoing[j].receiver == source:
                continue  # a transfer back to itself is not a fan-out edge
            group.append(outgoing[j])

        by_receiver: dict[str, list[Transaction]] = defaultdict(list)
        for tx in group:
            by_receiver[tx.receiver].append(tx)

        if len(by_receiver) >= config.min_intermediaries:
            yield by_receiver


def detect_fan_out_convergence(
    transactions: Sequence[Transaction],
    config: DetectorConfig | None = None,
) -> list[Finding]:
    """Detect fan-out -> convergence patterns in the given transactions.

    Callers doing replay must pre-filter `transactions` to what was
    observable at the evaluation time (see replay.evaluate_at) — this
    function has no notion of "now" and will use every transaction it is
    given, in either direction of time, as evidence.
    """
    config = config or DetectorConfig()
    txs = deduplicate_by_id(transactions)  # duplicate input records must not inflate evidence

    by_sender: dict[str, list[Transaction]] = defaultdict(list)
    for tx in txs:
        by_sender[tx.sender].append(tx)
    for sender_txs in by_sender.values():
        sender_txs.sort()

    findings: dict[tuple[str, str, tuple[str, ...]], Finding] = {}

    for source, outgoing in by_sender.items():
        for fan_out_by_intermediary in _fan_out_groups(source, outgoing, config):
            intermediaries = sorted(fan_out_by_intermediary)

            # For each intermediary, the fan-out transaction it is credited
            # with is the last one it received within this group (money must
            # have actually arrived before it can be forwarded on).
            fan_out_tx_for = {m: fan_out_by_intermediary[m][-1] for m in intermediaries}

            # collector -> intermediary -> earliest qualifying convergence tx
            convergence_candidates: dict[str, dict[str, Transaction]] = defaultdict(dict)
            for intermediary in intermediaries:
                received_at = fan_out_tx_for[intermediary].timestamp
                latest_allowed = received_at + datetime.timedelta(seconds=config.convergence_window_seconds)
                for tx in by_sender.get(intermediary, []):
                    if tx.timestamp <= received_at:
                        continue  # must be forwarded strictly after it was received
                    if tx.timestamp > latest_allowed:
                        continue
                    if tx.receiver == source or tx.receiver in fan_out_by_intermediary:
                        continue  # collector must be distinct from the source and the intermediaries
                    existing = convergence_candidates[tx.receiver].get(intermediary)
                    if existing is None or tx.timestamp < existing.timestamp:
                        convergence_candidates[tx.receiver][intermediary] = tx

            for collector, per_intermediary in convergence_candidates.items():
                if len(per_intermediary) < config.min_intermediaries:
                    continue

                converging = sorted(per_intermediary)
                fan_out_txs = [fan_out_tx_for[m] for m in converging]
                convergence_txs = [per_intermediary[m] for m in converging]

                window_start = min(t.timestamp for t in fan_out_txs)
                window_end = max(t.timestamp for t in convergence_txs)
                window_span_seconds = (window_end - window_start).total_seconds()

                total_fan_out_amount = sum(t.amount for t in fan_out_txs)
                total_convergence_amount = sum(t.amount for t in convergence_txs)

                intermediary_ratio = len(converging) / len(intermediaries)
                amount_conservation = (
                    min(1.0, total_convergence_amount / total_fan_out_amount) if total_fan_out_amount else 0.0
                )
                time_compactness = max(0.0, 1.0 - min(1.0, window_span_seconds / config.max_scoring_window_seconds))

                score = 0.5 * intermediary_ratio + 0.3 * amount_conservation + 0.2 * time_compactness
                score = round(min(1.0, max(0.0, score)), 4)

                key = (source, collector, tuple(converging))
                candidate = Finding(
                    pattern="fan_out_convergence",
                    source_account=source,
                    collector_account=collector,
                    intermediary_accounts=tuple(converging),
                    fan_out_transaction_ids=tuple(t.id for t in fan_out_txs),
                    convergence_transaction_ids=tuple(t.id for t in convergence_txs),
                    window_start=window_start,
                    window_end=window_end,
                    score=score,
                    score_method=_SCORE_METHOD,
                    evidence={
                        "intermediary_ratio": round(intermediary_ratio, 4),
                        "amount_conservation": round(amount_conservation, 4),
                        "time_compactness": round(time_compactness, 4),
                        "total_fan_out_amount": total_fan_out_amount,
                        "total_convergence_amount": total_convergence_amount,
                        "window_span_seconds": window_span_seconds,
                        "fan_out_window_seconds": config.fan_out_window_seconds,
                        "convergence_window_seconds": config.convergence_window_seconds,
                    },
                )

                existing = findings.get(key)
                if existing is None or candidate.score > existing.score:
                    findings[key] = candidate

    return sorted(findings.values(), key=lambda f: (-f.score, f.source_account, f.collector_account))

File: frontend/lib/services/rules/detector.ts
/**
 * Deterministic fan-out -> convergence detector.
 *
 * This is a line-for-line TypeScript port of ml/rules/detector.py, kept
 * intentionally faithful (same thresholds, same score formula, same
 * tie-breaking) so client-side replay produces exactly the same evidence
 * the backend would for the same observable transaction set. It has no
 * knowledge of any specific account id or the demo fixture -- every account
 * id it sees comes from the transactions passed in.
 *
 * The `score` this module produces is a hand-defined heuristic combining
 * three observable ratios. It is NOT a calibrated probability of fraud, NOT
 * a trained model's output, and NOT a measured performance metric -- treat
 * it as a heuristic ranking key, same as ml/rules/detector.py's docstring.
 */

import {
  compareTransactions,
  deduplicateById,
  formatUtcSeconds,
  type RuleTransaction,
} from "./transaction";

export interface DetectorConfig {
  /** Minimum distinct intermediaries required on both sides for a candidate group to be reported at all. */
  minIntermediaries: number;
  /** A fan-out group is a source account's outgoing transactions within this many seconds of the group's first transaction. */
  fanOutWindowSeconds: number;
  /** From when an intermediary receives a fan-out transaction, a subsequent outgoing transfer counts as convergence only within this many seconds. */
  convergenceWindowSeconds: number;
  /** Normalizes the time-compactness score term into [0, 1]; does not gate whether a finding is reported. */
  maxScoringWindowSeconds: number;
}

export const DEFAULT_DETECTOR_CONFIG: DetectorConfig = {
  minIntermediaries: 3,
  fanOutWindowSeconds: 60,
  convergenceWindowSeconds: 60,
  maxScoringWindowSeconds: 120,
};

export const SCORE_METHOD =
  "heuristic_v1 = 0.5*intermediary_ratio + 0.3*amount_conservation + 0.2*time_compactness, " +
  "clipped to [0, 1]. A hand-defined evidence-strength ranking score, not a calibrated " +
  "probability, not a trained model output, and not a measured performance metric.";

export interface Finding {
  pattern: "fan_out_convergence";
  sourceAccount: string;
  collectorAccount: string;
  intermediaryAccounts: string[];
  fanOutTransactionIds: string[];
  convergenceTransactionIds: string[];
  windowStart: string;
  windowEnd: string;
  score: number;
  scoreMethod: string;
  evidence: {
    intermediaryRatio: number;
    amountConservation: number;
    timeCompactness: number;
    totalFanOutAmount: number;
    totalConvergenceAmount: number;
    windowSpanSeconds: number;
    fanOutWindowSeconds: number;
    convergenceWindowSeconds: number;
  };
}

function round4(value: number): number {
  return Math.round(value * 10000) / 10000;
}

/**
 * Yields candidate fan-out groups for one source account. `outgoing` must
 * already be sorted by timestamp. Each group is anchored at one starting
 * transaction and extends forward while still inside the fan-out window, so
 * overlapping windows starting at different transactions are all considered
 * as separate candidates (deduplicated later by the findings map key).
 */
function* fanOutGroups(
  source: string,
  outgoing: RuleTransaction[],
  config: DetectorConfig,
): Generator<Map<string, RuleTransaction[]>> {
  const n = outgoing.length;
  for (let i = 0; i < n; i++) {
    const windowEndMs = outgoing[i].timestampMs + config.fanOutWindowSeconds * 1000;
    const group: RuleTransaction[] = [outgoing[i]];
    for (let j = i + 1; j < n; j++) {
      if (outgoing[j].timestampMs > windowEndMs) break;
      if (outgoing[j].receiver === source) continue; // a transfer back to itself is not a fan-out edge
      group.push(outgoing[j]);
    }

    const byReceiver = new Map<string, RuleTransaction[]>();
    for (const tx of group) {
      const list = byReceiver.get(tx.receiver);
      if (list) list.push(tx);
      else byReceiver.set(tx.receiver, [tx]);
    }

    if (byReceiver.size >= config.minIntermediaries) yield byReceiver;
  }
}

/**
 * Detects fan-out -> convergence patterns in the given transactions.
 * Callers doing replay must pre-filter `transactions` to what was
 * observable at the evaluation time (see replay.ts's evaluateAt) -- this
 * function has no notion of "now" and uses every transaction it is given.
 */
export function detectFanOutConvergence(
  transactions: readonly RuleTransaction[],
  config: DetectorConfig = DEFAULT_DETECTOR_CONFIG,
): Finding[] {
  const txs = deduplicateById(transactions); // duplicate input records must not inflate evidence

  const bySender = new Map<string, RuleTransaction[]>();
  for (const tx of txs) {
    const list = bySender.get(tx.sender);
    if (list) list.push(tx);
    else bySender.set(tx.sender, [tx]);
  }
  for (const list of bySender.values()) list.sort(compareTransactions);

  const findings = new Map<string, Finding>();

  for (const [source, outgoing] of bySender) {
    for (const fanOutByIntermediary of fanOutGroups(source, outgoing, config)) {
      const intermediaries = [...fanOutByIntermediary.keys()].sort();

      // For each intermediary, the fan-out transaction it is credited with
      // is the last one it received within this group (money must have
      // actually arrived before it can be forwarded on).
      const fanOutTxFor = new Map<string, RuleTransaction>();
      for (const m of intermediaries) {
        const list = fanOutByIntermediary.get(m)!;
        fanOutTxFor.set(m, list[list.length - 1]);
      }

      // collector -> intermediary -> earliest qualifying convergence tx
      const convergenceCandidates = new Map<string, Map<string, RuleTransaction>>();
      for (const intermediary of intermediaries) {
        const receivedAtMs = fanOutTxFor.get(intermediary)!.timestampMs;
        const latestAllowedMs = receivedAtMs + config.convergenceWindowSeconds * 1000;
        const senderTxs = bySender.get(intermediary) ?? [];
        for (const tx of senderTxs) {
          if (tx.timestampMs <= receivedAtMs) continue; // must be forwarded strictly after it was received
          if (tx.timestampMs > latestAllowedMs) continue;
          if (tx.receiver === source || fanOutByIntermediary.has(tx.receiver)) continue; // collector must be distinct from source and intermediaries

          let perIntermediary = convergenceCandidates.get(tx.receiver);
          if (!perIntermediary) {
            perIntermediary = new Map();
            convergenceCandidates.set(tx.receiver, perIntermediary);
          }
          const existing = perIntermediary.get(intermediary);
          if (!existing || tx.timestampMs < existing.timestampMs) {
            perIntermediary.set(intermediary, tx);
          }
        }
      }

      for (const [collector, perIntermediary] of convergenceCandidates) {
        if (perIntermediary.size < config.minIntermediaries) continue;

        const converging = [...perIntermediary.keys()].sort();
        const fanOutTxs = converging.map((m) => fanOutTxFor.get(m)!);
        const convergenceTxs = converging.map((m) => perIntermediary.get(m)!);

        const windowStartMs = Math.min(...fanOutTxs.map((t) => t.timestampMs));
        const windowEndMs = Math.max(...convergenceTxs.map((t) => t.timestampMs));
        const windowSpanSeconds = (windowEndMs - windowStartMs) / 1000;

        const totalFanOutAmount = fanOutTxs.reduce((sum, t) => sum + t.amount, 0);
        const totalConvergenceAmount = convergenceTxs.reduce((sum, t) => sum + t.amount, 0);

        const intermediaryRatio = converging.length / intermediaries.length;
        const amountConservation = totalFanOutAmount
          ? Math.min(1, totalConvergenceAmount / totalFanOutAmount)
          : 0;
        const timeCompactness = Math.max(
          0,
          1 - Math.min(1, windowSpanSeconds / config.maxScoringWindowSeconds),
        );

        let score = 0.5 * intermediaryRatio + 0.3 * amountConservation + 0.2 * timeCompactness;
        score = round4(Math.min(1, Math.max(0, score)));

        const key = `${source}|${collector}|${converging.join(",")}`;
        const candidate: Finding = {
          pattern: "fan_out_convergence",
          sourceAccount: source,
          collectorAccount: collector,
          intermediaryAccounts: converging,
          fanOutTransactionIds: fanOutTxs.map((t) => t.id),
          convergenceTransactionIds: convergenceTxs.map((t) => t.id),
          windowStart: formatUtcSeconds(windowStartMs),
          windowEnd: formatUtcSeconds(windowEndMs),
          score,
          scoreMethod: SCORE_METHOD,
          evidence: {
            intermediaryRatio: round4(intermediaryRatio),
            amountConservation: round4(amountConservation),
            timeCompactness: round4(timeCompactness),
            totalFanOutAmount,
            totalConvergenceAmount,
            windowSpanSeconds,
            fanOutWindowSeconds: config.fanOutWindowSeconds,
            convergenceWindowSeconds: config.convergenceWindowSeconds,
          },
        };

        const existing = findings.get(key);
        if (!existing || candidate.score > existing.score) {
          findings.set(key, candidate);
        }
      }
    }
  }

  return [...findings.values()].sort((a, b) => {
    if (a.score !== b.score) return b.score - a.score;
    if (a.sourceAccount !== b.sourceAccount) return a.sourceAccount < b.sourceAccount ? -1 : 1;
    if (a.collectorAccount !== b.collectorAccount) return a.collectorAccount < b.collectorAccount ? -1 : 1;
    return 0;
  });
}
